src/output.rs

```diff
@@ -440,3 +440,51 @@ impl PartialEq<Output> for WeakOutput {
         self.upgrade().map(|o| &o == other).unwrap_or(false)
     }
 }
+
+/// Source for determining output mode information.
+#[derive(PartialEq, Clone, Debug)]
+pub enum OutputModeSource {
+    /// Automatic mode based on an [`Output`].
+    Auto(Output),
+    /// Static output mode.
+    Static {
+        /// Size of the static output
+        size: Size<i32, Physical>,
+        /// Scale of the static output
+        scale: utils::Scale<f64>,
+        /// Transform of the static output
+        transform: Transform,
+    },
+}
+
+impl TryFrom<&OutputModeSource> for (Size<i32, Physical>, utils::Scale<f64>, Transform) {
+    type Error = OutputNoMode;
+
+    fn try_from(mode: &OutputModeSource) -> Result<Self, Self::Error> {
+        match mode {
+            OutputModeSource::Auto(output) => Ok((
+                output.current_mode().ok_or(OutputNoMode)?.size,
+                output.current_scale().fractional_scale().into(),
+                output.current_transform(),
+            )),
+            OutputModeSource::Static {
+                size,
+                scale,
+                transform,
+            } => Ok((*size, *scale, *transform)),
+        }
+    }
+}
+
+impl TryFrom<OutputModeSource> for (Size<i32, Physical>, utils::Scale<f64>, Transform) {
```

Comment:
Switched this to `TryFrom`, because it's better and I don't know why it wasn't `TryFrom` to begin with.